### PR TITLE
Feature/update conditions for pswap lp strategic bonus rewards

### PR DESCRIPTION
**Pull request by HaidarJbeily7**

### CODE REVIEW

File: pallets/farming/src/lib.rs [hunk 1]
```diff
@@ -388,6 +398,29 @@ pub mod pallet {
...
+        #[pallet::call_index(0)]
+        #[pallet::weight(<T as Config>::WeightInfo::set_lp_min_xor_for_bonus_reward())]
+        pub fn set_lp_min_xor_for_bonus_reward(
+            origin: OriginFor<T>,
+            new_lp_min_xor_for_bonus_reward: Balance,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            LpMinXorForBonusReward::<T>::mutate(|lp_min_xor_for_bonus_reward| {
+                *lp_min_xor_for_bonus_reward = new_lp_min_xor_for_bonus_reward;
+            });
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:417]:
```suggestion
            LpMinXorForBonusReward::<T>::set(new_lp_min_xor_for_bonus_reward);
```

Comment (wer1st) [reply pallets/farming/src/lib.rs:417]: or `put`

File: pallets/farming/src/lib.rs [hunk 2]
```diff
@@ -388,6 +398,29 @@ pub mod pallet {
...
+    #[pallet::getter(fn lp_min_xor_for_bonus_reward)]
+    pub type LpMinXorForBonusReward<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(<T as Config>::WeightInfo::set_lp_min_xor_for_bonus_reward())]
+        pub fn set_lp_min_xor_for_bonus_reward(
+            origin: OriginFor<T>,
+            new_lp_min_xor_for_bonus_reward: Balance,
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:412]: Do we need to check `new_lp_min_xor_for_bonus_reward` value somehow? Is `new_lp_min_xor_for_bonus_reward = 0` ok?

File: pallets/farming/src/lib.rs [hunk 3]
```diff
@@ -185,7 +185,7 @@ impl<T: Config> Pallet<T> {
             .try_into_balance()
             .unwrap_or(0);
 
-        if base_asset_amt < balance!(1) {
+        if base_asset_amt < Self::lp_min_xor_for_bonus_reward() {
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:188]: If the value < `lp_min_xor_for_bonus_reward` the 0 is returned as a weight. It means that if this value is too high, liquidity providers will not get any reward. @vovac12 is it intended?

File: pallets/farming/src/lib.rs [hunk 4]
```diff
@@ -379,6 +381,14 @@ pub mod pallet {
         IncRefError,
     }
 
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        LpMinXorForBonusRewardUpdated {
+            lp_min_xor_for_bonus_reward: Balance,
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:388]: Add the old value as well

File: pallets/farming/src/lib.rs [hunk 5]
```diff
@@ -379,6 +381,14 @@ pub mod pallet {
         IncRefError,
     }
 
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        LpMinXorForBonusRewardUpdated {
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:387]: Add the doc comment `///` for the event

File: pallets/farming/src/lib.rs [hunk 6]
```diff
@@ -388,6 +398,29 @@ pub mod pallet {
     #[pallet::storage]
     pub type PoolFarmers<T: Config> =
         StorageMap<_, Identity, T::AccountId, Vec<PoolFarmer<T>>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn lp_min_xor_for_bonus_reward)]
+    pub type LpMinXorForBonusReward<T: Config> = StorageValue<_, Balance, ValueQuery>;
```
Comment (wer1st) [inline pallets/farming/src/lib.rs:404]: You have to add `DefaultLpMinXorForBonusReward` value

Comment (wer1st) [reply pallets/farming/src/lib.rs:404]: And, probably add the migration to inroduce the first value. @vovac12 what do you think?
And what about adding the default value in the chain spec?

File: pallets/farming/src/weights.rs [hunk 7]
```diff
@@ -63,6 +63,7 @@ pub trait WeightInfo {
 	fn refresh_pool(a: u32, ) -> Weight;
 	fn prepare_accounts_for_vesting(a: u32, b: u32, ) -> Weight;
 	fn vest_account_rewards(a: u32, ) -> Weight;
+	fn set_lp_min_xor_for_bonus_reward() -> Weight;
```
Comment (wer1st) [inline pallets/farming/src/weights.rs:66]: I don't see the benchmarks. You need to add the benchmark of the extrinsic in the `benchmarking.rs` and modiy the `WeightInfo` inside the `farming::Config` in the `runtime\src\lib.rs`

File: pallets/farming/src/lib.rs (hunk 2 above)
Comment (HaidarJbeily7) [reply pallets/farming/src/lib.rs:422]: I added a check with a subjective error

File: pallets/farming/src/lib.rs (hunk 5 above)
Comment (HaidarJbeily7) [reply pallets/farming/src/lib.rs:390]: added

File: pallets/farming/src/lib.rs (hunk 4 above)
Comment (HaidarJbeily7) [reply pallets/farming/src/lib.rs]: added

File: pallets/farming/src/lib.rs (hunk 6 above)
Comment (vovac12) [reply pallets/farming/src/lib.rs]: It's enough to add default value

File: pallets/farming/src/lib.rs (hunk 3 above)
Comment (vovac12) [reply pallets/farming/src/lib.rs:188]: Yes, that's why we have this condition

File: pallets/farming/src/lib.rs [hunk 8]
```diff
@@ -377,6 +379,23 @@ pub mod pallet {
...
+        /// When Minimum XOR amount for Liquidity Provider Bonus Reward is updated
+        LpMinXorForBonusRewardUpdated {
+            new_lp_min_xor_for_bonus_reward: Balance,
+            old_lp_min_xor_for_bonus_reward: Balance,
+        },
+    }
+
+    #[pallet::type_value]
+    pub fn DefaultLpMinXorForBonusReward<T: Config>() -> Balance {
+        balance!(1)
```
Comment (vovac12) [inline pallets/farming/src/lib.rs:398]: I think it's better to set default value to one mentioned in issue to apply changes right after the next runtime upgrade

File: pallets/farming/src/lib.rs [hunk 9]
```diff
@@ -388,6 +407,34 @@ pub mod pallet {
...
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(<T as Config>::WeightInfo::set_lp_min_xor_for_bonus_reward())]
+        pub fn set_lp_min_xor_for_bonus_reward(
+            origin: OriginFor<T>,
+            new_lp_min_xor_for_bonus_reward: Balance,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            ensure!(
+                new_lp_min_xor_for_bonus_reward >= balance!(1),
```
Comment (vovac12) [inline pallets/farming/src/lib.rs:426]: Balance is u128 which is always >= 0

File: runtime/src/lib.rs [hunk 10]
```diff
@@ -2410,7 +2411,7 @@ construct_runtime! {
         ElectionsPhragmen: pallet_elections_phragmen::{Pallet, Call, Storage, Event<T>, Config<T>} = 39,
         VestedRewards: vested_rewards::{Pallet, Call, Storage, Event<T>} = 40,
         Identity: pallet_identity::{Pallet, Call, Storage, Event<T>} = 41,
-        Farming: farming::{Pallet, Storage} = 42,
+        Farming: farming::{Pallet, Storage, Event<T>} = 42,
```
Comment (vovac12) [inline runtime/src/lib.rs:2414]: Also add `Call` here

File: pallets/farming/src/lib.rs (hunk 9 above)
Comment (wer1st) [reply pallets/farming/src/lib.rs:426]: Ok, I clarified just to not miss the error.

File: pallets/farming/src/benchmarking.rs [hunk 11]
```diff
@@ -160,6 +160,14 @@ benchmarks! {
     }: {
         Pallet::<T>::vest_account_rewards(accounts);
     }
+
+    set_lp_min_xor_for_bonus_reward{
+        let root = RawOrigin::Root;
+        let balance = balance!(10000000);
+    }: {
+       let _ = Pallet::<T>::set_lp_min_xor_for_bonus_reward(root.into(), balance);
```
Comment (wer1st) [inline pallets/farming/src/benchmarking.rs:168]:
```suggestion
       Pallet::<T>::set_lp_min_xor_for_bonus_reward(root.into(), balance).unwrap();
```

Comment (wer1st) [reply pallets/farming/src/benchmarking.rs:168]: It allows us to get panic if the code returns the error. Just `let _ = ...` is ok with the error result.

File: pallets/farming/src/tests.rs [hunk 12]
```diff
@@ -63,6 +65,8 @@ fn init_pool(dex_id: DEXId, base_asset: AssetId, other_asset: AssetId) {
 fn test() {
     let dex_id = DEX_A_ID;
     ExtBuilder::default().build().execute_with(|| {
+        let _ =
+            <Pallet<Runtime>>::set_lp_min_xor_for_bonus_reward(RawOrigin::Root.into(), balance!(1));
```
Comment (wer1st) [inline pallets/farming/src/tests.rs:69]:
```suggestion
        <Pallet<Runtime>>::set_lp_min_xor_for_bonus_reward(RawOrigin::Root.into(), balance!(1)).unwrap();
```

Comment (wer1st) [inline pallets/farming/src/tests.rs:69]: or
```suggestion
        assert_ok!(<Pallet<Runtime>>::set_lp_min_xor_for_bonus_reward(RawOrigin::Root.into(), balance!(1)));
```

Comment (wer1st) [inline pallets/farming/src/tests.rs:69]: Probably you can also try to remove this line and test the default value.

**Review by wer1st — APPROVED** [review]

**Review by vovac12 — APPROVED** [review]

File: pallets/farming/src/tests.rs [hunk 13]
```diff
@@ -406,3 +423,28 @@ fn test() {
...
+
+#[test]
+fn set_lp_min_xor_for_bonus_reward_should_work() {
+    ExtBuilder::default().build().execute_with(|| {
+        let modified_min_xor = balance!(3 * (10_i32.pow(6)));
+        assert_ok!(Pallet::<Runtime>::set_lp_min_xor_for_bonus_reward(
+            RawOrigin::Root.into(),
+            modified_min_xor
+        ));
+        assert!(Pallet::<Runtime>::lp_min_xor_for_bonus_reward() == modified_min_xor);
```
Comment (Alexey-N-Chernyshov) [inline pallets/farming/src/tests.rs:448]: It would be nice to test that the event was emitted with `assert_has_event`.

Comment (HaidarJbeily7) [reply pallets/farming/src/tests.rs]: added assert for event emitting.
@Alexey-N-Chernyshov

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**